In [1]:
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [2]:
# Set the path to the data
excel_path = "data/c14_master_v08.xlsx"

In [3]:
df = pd.read_excel(excel_path)
df.head()

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,pMC_value,pMC_error,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
0,1,Jörlanda,Bohuslän,Kyrkeby 3:34 m.fl.,Jörlanda 158,Boplats,L1969:5267,103367.0,Ua-49252,2841.0,33.0,25.3,NaN,NaN,ok,NaN,J,Träkol,Lind,A1328,Kokgrop,6432029,312382,11.826064,57.990215,archaeological site,-930.0,-1046.0,-913.0,-1110.0,-997.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Åberg, Joakim",2015.0,Bergsmonument och boplatser i Jörlanda,"Bohusläns museum, Rapport 2015:14",Uddevalla
1,2,Jörlanda,Bohuslän,Berg1:69,Jörlanda 185,Boplats,L1970:9431,NaN,T-8772,1720.0,80.0,NaN,NaN,NaN,ok,NaN,J,Träkol,NaN,A4,Stenpackning,6435710,313378,11.839973,58.023646,archaeological site,414.0,247.0,541.0,132.0,339.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Johansson, Nils",1995.0,Tre boplatser i Spekerödsdalen,"UV Väst, Rapport 1995:1",Kungsbacka
2,3,Jörlanda,Bohuslän,Berg1:69,Jörlanda 185,Boplats,L1970:9431,NaN,T-8773,2890.0,105.0,NaN,NaN,NaN,ok,NaN,J,Träkol,NaN,A6,Stenpackning,6435710,313378,11.839973,58.023646,archaeological site,-930.0,-1214.0,-830.0,-1381.0,-1086.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Johansson, Nils",1995.0,Tre boplatser i Spekerödsdalen,"UV Väst, Rapport 1995:1",Kungsbacka
3,4,Jörlanda,Bohuslän,Berg1:69,Jörlanda 185,Boplats,L1970:9431,NaN,T-8774,2015.0,75.0,NaN,NaN,NaN,ok,NaN,J,Träkol,NaN,A8,Härd,6435710,313378,11.839973,58.023646,archaeological site,117.0,-101.0,207.0,-333.0,-5.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Johansson, Nils",1995.0,Tre boplatser i Spekerödsdalen,"UV Väst, Rapport 1995:1",Kungsbacka
4,5,Jörlanda,Bohuslän,Berg1:69,Jörlanda 185,Boplats,L1970:9431,NaN,T-8775,1990.0,55.0,NaN,NaN,NaN,ok,NaN,J,Träkol,NaN,A23,Grop,6435710,313378,11.839973,58.023646,archaeological site,110.0,-42.0,203.0,-144.0,29.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Johansson, Nils",1995.0,Tre boplatser i Spekerödsdalen,"UV Väst, Rapport 1995:1",Kungsbacka


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 30301 entries, 0 to 30300
Data columns (total 38 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fid                   30301 non-null  int64  
 1   socken                30301 non-null  str    
 2   landskap              30301 non-null  str    
 3   village_farm          30244 non-null  str    
 4   raa_id                28460 non-null  str    
 5   site_type             30219 non-null  str    
 6   site_id               29963 non-null  str    
 7   uppdragsnummer        17681 non-null  float64
 8   lab_no                30301 non-null  str    
 9   c14_age_bp            30185 non-null  float64
 10  c14_error             30152 non-null  float64
 11  d13C                  21439 non-null  float64
 12  pMC_value             100 non-null    float64
 13  pMC_error             99 non-null     float64
 14  c14_data_status       30301 non-null  str    
 15  comment               3682 non

In [5]:
df.describe(include='all')

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,pMC_value,pMC_error,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
count,30301.00000,30301,30301,30244,28460,30219,29963,1.768100e+04,30301,30185.000000,30152.000000,21439.000000,100.000000,99.000000,30301,3682,30106,30300,18240,26035,29795,3.030100e+04,3.030100e+04,30301.000000,30301.000000,30301,29731.000000,29731.000000,29731.000000,29731.000000,29731.000000,29731,30301,30301,30269.000000,30277,29833,29605
unique,NaN,1056,37,4998,5437,303,6118,NaN,30291,NaN,NaN,NaN,NaN,NaN,3,2058,4,41,583,10503,1053,NaN,NaN,NaN,NaN,2,NaN,NaN,NaN,NaN,NaN,1,3,2020,NaN,4498,4312,140
top,NaN,Uppsala,Uppland,Lockarp 2:1 m fl,Lockarp 18,Boplats,L1988:4693,NaN,Ua-86040,NaN,NaN,NaN,NaN,NaN,ok,Daterar ej anläggningen,J,Träkol,Tall,A1,Härd,NaN,NaN,NaN,NaN,archaeological site,NaN,NaN,NaN,NaN,NaN,HPD; IntCal20; rcarbon 1.5.0,calibrated,Protokoll,NaN,C-14 Dateringar,Stencil,Stockholm
freq,NaN,1026,6042,223,223,8449,223,NaN,3,NaN,NaN,NaN,NaN,NaN,29731,188,28364,21891,2928,718,6966,NaN,NaN,NaN,NaN,29963,NaN,NaN,NaN,NaN,NaN,29731,29731,485,NaN,485,546,6401
mean,15151.00000,NaN,NaN,NaN,NaN,NaN,NaN,1.739642e+08,NaN,2140.700977,54.636973,20.249556,113.966000,0.979293,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.486949e+06,5.128824e+05,15.303081,58.502299,NaN,-176.953180,-359.532811,-118.680334,-419.985032,-266.149810,NaN,NaN,NaN,1967.210810,NaN,NaN,NaN
std,8747.28959,NaN,NaN,NaN,NaN,NaN,NaN,8.004354e+07,NaN,2033.402101,82.896531,190.911893,30.514774,3.138648,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.255938e+05,1.328164e+05,2.285725,2.020949,NaN,1996.170029,2026.434700,1981.253296,2048.037816,2008.703843,NaN,NaN,NaN,354.604782,NaN,NaN,NaN
min,1.00000,NaN,NaN,NaN,NaN,NaN,NaN,1.000180e+05,NaN,-1.000000,-1.000000,-28.300000,80.900000,0.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.653320e+05,2.788610e+05,11.154336,4.209866,NaN,-44302.000000,-45589.000000,-43964.000000,-46225.000000,-45016.000000,NaN,NaN,NaN,0.000000,NaN,NaN,NaN
25%,7576.00000,NaN,NaN,NaN,NaN,NaN,NaN,2.004001e+08,NaN,1140.000000,30.000000,0.000000,99.997500,0.400000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.311431e+06,3.818210e+05,13.103535,56.924534,NaN,-545.000000,-777.000000,-421.000000,-798.000000,-636.000000,NaN,NaN,NaN,1998.000000,NaN,NaN,NaN
50%,15151.00000,NaN,NaN,NaN,NaN,NaN,NaN,2.012007e+08,NaN,1801.000000,40.000000,24.600000,103.105000,0.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.501630e+06,5.126990e+05,15.219477,58.631743,NaN,325.000000,132.000000,375.000000,85.000000,233.000000,NaN,NaN,NaN,2007.000000,NaN,NaN,NaN
75%,22726.00000,NaN,NaN,NaN,NaN,NaN,NaN,2.018011e+08,NaN,2505.000000,70.000000,26.200000,116.450000,0.700000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.606790e+06,6.386760e+05,17.456184,59.576556,NaN,973.000000,775.000000,992.000000,691.000000,860.500000,NaN,NaN,NaN,2015.000000,NaN,NaN,NaN


In [6]:
df.dropna(subset='species')['material'].unique()

<StringArray>
[                'Träkol',            'Ben, brända',          'Förkolnat frö',
              'Växtdelar',                   'Tand',    'Organisk beläggning',
                    'Trä',           'Ben, obrända',           'Skalfragment',
              'Animalier',                   'Bark',     'Frö (ej förkolnat)',
                   'Horn',               'Fekalier',               'Kollagen',
                  'Svamp',             'Kol i järn',     'Organiskt material',
 'Sot/Organiskt material',                  'Harts',              'Järnslagg',
                  'Okänd',                   'Bröd',               'Karbonat',
              'Förkolnad',         'Skal/mollusker',               'Ben,horn',
        'Krossad keramik',                   'Fisk']
Length: 29, dtype: str

In [7]:
#pMC values

df_pmc = df.dropna(subset='pMC_value')
df_pmc

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,pMC_value,pMC_error,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
74,75,Boo,Uppland,"Boo 1:30, 11:3",Boo 22,Tegelindustri,L2017:1373,201300088.0,Ua-40545,NaN,NaN,25.0,104.5,0.5,c14_data_saknas,NaN,N,Mossa,NaN,L2,Lager,6579309,687654,18.296883,59.310833,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Ljung, J-Å, Strucke, Ulf & Vinberg, Ann",2013.0,Tegelugnen i Boo. Medeltida tegeltillverkning ...,UV Rapport 2013:27,Stockholm
751,752,Växjö,Småland,Domkyrkan 1,Växjö 170,Stadslager,L1953:1681,201700034.0,Ua-54631,NaN,NaN,NaN,101.6,0.5,c14_data_saknas,NaN,N,Träkol,NaN,R31,Röjningsröse,6303851,488383,14.809389,56.878263,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Bondesson Hvid, Bo",2017.0,"Arkeologisk förundersökning 2016, Växjö 13:32,...","CMB Uppdragsarkeologi, 2017:15",Löddeköppinge
835,836,Karlstad,Värmland,Kv Druvan,Karlstad 30,Stadslager,L2007:9898,200400205.0,Ua-27875,0.0,0.0,25.3,99.7,0.4,c14_data_saknas,NaN,J,Träkol,Björk,8369,Stolphål,6583565,414665,13.497923,59.382118,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Karlsson, Martin & Knabe, Ebba",2006.0,"Stadsarkeologi i kvarteret Druvan, Karlstad","UV Bergslagen, Dokumentation av fältarbetsfase...",Stockholm
1223,1224,Björlanda,Bohuslän,Kvisljungeby 2:122,Björlanda 303,Boplatsområde,L1970:6811,201900585.0,Ua-63437,NaN,NaN,24.1,106.1,0.4,c14_data_saknas,Daterar ej anläggningen,N,Träkol,Al,A1000,Grop,6405020,310885,11.822228,57.747355,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Beckman, Anna",2020.0,En boplats i Björlanda,"Kulturmiljö, Förvaltningen för kulturutvecklin...",Vänersborg
1608,1609,Gällivare,Lappland,Sakajärvi 2:4,NaN,Härd,L2018:327,201800424.0,Ua-62165,NaN,NaN,13.4,101.6,0.6,c14_data_saknas,Daterar ej anläggningen,N,"Ben, obrända",NaN,NaN,Härd,7450181,751891,20.798940,67.063229,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Lindgren, Åsa",2018.0,"Aitik, en ny härd 2018","Norrbottens museum, Rapport 2019:2",Luleå
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29676,29677,Västerås,Västmanland,Stadsparken,Västerås 232,Stadslager,L2002:434,201801994.0,Ua-55311,NaN,NaN,26.8,113.5,0.0,c14_data_saknas,NaN,N,Kol/sot,NaN,S4,Lager,6609051,587059,16.542816,59.610546,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Spjutj, Oskar",2017.0,Vikingatida kulturlager: Upprustning av Stadsp...,"Stiftelsen Kulturmiljövård, Rapport 2017:41",Västerås
30026,30027,Lindome,Ha,Inseros 1:3,287,Fast fiske,L1969:4490,103330.0,Ua-47703,NaN,NaN,29.6,99.3,0.4,c14_data_saknas,NaN,J,Trä,Björk,NaN,Påle,6388296,336143,12.257408,57.607233,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Bergstrand, Thomas",2014.0,Inseros. Fasta fiskeanläggningar i Västra Ingsjön,"Bohusläns museum, Rapport 2014:9",Uddevalla
30027,30028,Lindome,Ha,Inseros 1:3,366,Fast fiske,L1959:973,103330.0,Ua-47702,NaN,NaN,28.1,99.9,0.4,c14_data_saknas,NaN,J,Trä,barrträd,NaN,Påle,6388269,336147,12.257493,57.606992,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Bergstrand, Thomas",2014.0,Inseros. Fasta fiskeanläggningar i Västra Ingsjön,"Bohusläns museum, Rapport 2014:9",Uddevalla
30063,30064,Mistelås,Sm,Villarp 1:7,89,Fossil åker,L1954:9018,201901235.0,Ua-65477,NaN,NaN,27.0,106.9,0.4,c14_data_saknas,NaN,J,Träkol,NaN,RR37,Röjningsröse,6312508,463485,14.399633,56.954738,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Engman, Fredrik",2020.0,Arkeologisk förundersökning. Del av fornlämnin...,SWECO RAPPORT Uppdragsnummer 13009305,Stockh

In [8]:
# save pMC data (all instances with pMC_value being non-null)
# we will assume it is a subset of data where a measurement was done, unlike the rest of the data where this value is null

df_pmc.to_csv('output/strucke_v8_pMC_values.csv')

In [9]:
#remaining

df_no_pmc = df[df['pMC_value'].isna()].drop(columns=['pMC_value', 'pMC_error'])
df_no_pmc.describe(include='all')

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
count,30201.000000,30201,30201,30145,28372,30120,29869,1.760600e+04,30201,30158.000000,30124.000000,21373.000000,30201,3655,30006,30200,18177,25953,29696,3.020100e+04,3.020100e+04,30201.000000,30201.000000,30201,29730.000000,29730.000000,29730.000000,29730.000000,29730.000000,29730,30201,30201,30169.000000,30177,29733,29506
unique,NaN,1054,37,4990,5427,303,6106,NaN,30191,NaN,NaN,NaN,3,2046,4,39,581,10480,1048,NaN,NaN,NaN,NaN,2,NaN,NaN,NaN,NaN,NaN,1,3,2019,NaN,4490,4304,140
top,NaN,Uppsala,Uppland,Lockarp 2:1 m fl,Lockarp 18,Boplats,L1988:4693,NaN,Ua-86040,NaN,NaN,NaN,ok,Daterar ej anläggningen,J,Träkol,Tall,A1,Härd,NaN,NaN,NaN,NaN,archaeological site,NaN,NaN,NaN,NaN,NaN,HPD; IntCal20; rcarbon 1.5.0,calibrated,Protokoll,NaN,C-14 Dateringar,Stencil,Stockholm
freq,NaN,1024,6030,223,223,8429,223,NaN,3,NaN,NaN,NaN,29730,176,28322,21841,2920,712,6954,NaN,NaN,NaN,NaN,29869,NaN,NaN,NaN,NaN,NaN,29730,29730,485,NaN,485,546,6380
mean,15147.466177,NaN,NaN,NaN,NaN,NaN,NaN,1.738358e+08,NaN,2142.600438,54.685035,20.246425,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.486599e+06,5.128228e+05,15.301778,58.499172,NaN,-177.007770,-359.589573,-118.733838,-420.043054,-266.206256,NaN,NaN,NaN,1967.065564,NaN,NaN,NaN
std,8750.229094,NaN,NaN,NaN,NaN,NaN,NaN,7.889315e+07,NaN,2033.318427,82.919346,191.205376,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.253200e+05,1.327954e+05,2.284450,2.018586,NaN,1996.181408,2026.445146,1981.265138,2048.047824,2008.714047,NaN,NaN,NaN,355.182724,NaN,NaN,NaN
min,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,1.000180e+05,NaN,-1.000000,-1.000000,-28.300000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.653320e+05,2.788610e+05,11.154336,4.209866,NaN,-44302.000000,-45589.000000,-43964.000000,-46225.000000,-45016.000000,NaN,NaN,NaN,0.000000,NaN,NaN,NaN
25%,7566.000000,NaN,NaN,NaN,NaN,NaN,NaN,2.004001e+08,NaN,1140.000000,30.000000,0.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.311270e+06,3.817780e+05,13.102906,56.923493,NaN,-545.000000,-777.000000,-421.000000,-798.000000,-636.000000,NaN,NaN,NaN,1998.000000,NaN,NaN,NaN
50%,15144.000000,NaN,NaN,NaN,NaN,NaN,NaN,2.012007e+08,NaN,1803.500000,40.000000,24.600000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.501320e+06,5.126990e+05,15.219477,58.631743,NaN,325.000000,132.000000,375.000000,85.000000,233.000000,NaN,NaN,NaN,2007.000000,NaN,NaN,NaN
75%,22733.000000,NaN,NaN,NaN,NaN,NaN,NaN,2.018011e+08,NaN,2505.000000,70.000000,26.200000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.606722e+06,6.385790e+05,17.456184,59.575896,NaN,973.000000,775.000000,992.000000,691.000000,860.000000,NaN,NaN,NaN,2015.000000,NaN,NaN,NaN


In [10]:
df_c14_data_saknas = df_no_pmc[df_no_pmc['c14_data_status']=='c14_data_saknas']
df_c14_data_saknas

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
100,101,Våmb,Västergötland,Våmb 30:10,Våmb 52,Område med fossil åkermark,L1961:6180,201600692.0,ICA-Okänd 3,0.0,40.0,NaN,c14_data_saknas,NaN,J,Träkol,NaN,A114,Lager,6473028,429916,13.801147,58.392444,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Berger, Åsa",2016.0,Stensättningar och järnframställning - Arkeolo...,"Arkeologikonsult, Rapport 2016:2941",Upplands Väsby
344,345,Husby-Ärlinghundra,Uppland,"Måby 2:7 ""Maren""",Husby-Ärlinghundra 138,Lägenhetsbebyggelse,L2016:1375,NaN,ST13802,250.0,0.0,0.0,c14_data_saknas,NaN,J,Träkol,NaN,A 264,Stolphål,6613144,664703,17.920385,59.623851,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Hallesjö, Maria & Hamilton, John",1997.0,En boplats och ett torp vid Maren,"UV Stockholm, Rapport 1997:33",Stockholm
345,346,Husby-Ärlinghundra,Uppland,"Måby 2:7 ""Maren""",Husby-Ärlinghundra 138,Lägenhetsbebyggelse,L2016:1375,NaN,ST13803,250.0,0.0,0.0,c14_data_saknas,NaN,J,Träkol,NaN,A 711,Grop,6613144,664703,17.920385,59.623851,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Hallesjö, Maria & Hamilton, John",1997.0,En boplats och ett torp vid Maren,"UV Stockholm, Rapport 1997:33",Stockholm
346,347,Husby-Ärlinghundra,Uppland,"Måby 2:7 ""Maren""",Husby-Ärlinghundra 138,Lägenhetsbebyggelse,L2016:1375,NaN,ST13804,250.0,0.0,0.0,c14_data_saknas,NaN,J,Träkol,NaN,A 826,Grop,6613144,664703,17.920385,59.623851,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Hallesjö, Maria & Hamilton, John",1997.0,En boplats och ett torp vid Maren,"UV Stockholm, Rapport 1997:33",Stockholm
644,645,Växjö,Småland,Biskopshagen,Växjö 358,Grav- och boplatsområde,L1953:8423,NaN,PozOkänd 11,NaN,NaN,NaN,c14_data_saknas,Labnr och 14Cålder saknas i rapport Poz? 1460-...,NaN,Träkol,NaN,A3829,Stolphål,6302567,487812,14.800081,56.866713,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kadefors, Ola",2006.0,Boplatsen i Kvarteren Biskopshagen,"Smålands museum, Rapport 2006:51",Växjö
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
30250,30251,Tossene,Bo,Torp 1:13,1018,Boplats,L1960:8757,202500158.0,Poz-40188,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i rapport. Kal. 2 sigma 1375-...,J,"Ben, brända",NaN,A16735,Kokgrop,6484551,288035,11.367227,58.450088,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kegel, K., Sörman, A., Wessman, A. & Ling, J.",2013.0,Arkeologisk undersökning av RAÄ 1018 vid hällr...,GOTARC serie D Arkeologiska rapporter no 99.,Göteborg
30251,30252,Tossene,Bo,Torp 1:13,1018,Boplats,L1960:8757,202500158.0,Poz-40189,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i rapport kal. 2 sigma 1730-1...,J,"Ben, brända",NaN,A21351,Stolphål,6484551,288035,11.367227,58.450088,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kegel, K., Sörman, A., Wessman, A. & Ling, J.",2013.0,Arkeologisk undersökning av RAÄ 1018 vid hällr...,GOTARC serie D Arkeologiska rapporter no 99.,Göteborg
30262,30263,Tossene,Bo,Torp 1:13,1018,Boplats,L1960:8757,202500158.0,Ua-40187,NaN,NaN,NaN,c14_data_saknas,Labnr saknas i rapport Kal. 2 sigma 1680-1515 BC,J,"Ben, brända",NaN,A17128,Kokgrop/ugn,6484551,288035,11.367227,58.450088,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kegel, K., Sörman, A., Wessman, A. & Ling, J.",2013.0,Arkeologisk undersökning av RAÄ 1018 vid hällr...,GOTARC serie D Arkeologiska rapporter no 99.,Göteborg
30265,30266,Tossene,Bo,Torp 1:13,1018,Boplats,L1960:8757,202500158.0,Ua-44252,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i r

In [11]:
#values with no data for c14 nor cal(68 nor 95), where only material, species and author exist
df_c14_data_saknas[df_c14_data_saknas['c14_age_bp'].isna()]

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
644,645,Växjö,Småland,Biskopshagen,Växjö 358,Grav- och boplatsområde,L1953:8423,NaN,PozOkänd 11,NaN,NaN,NaN,c14_data_saknas,Labnr och 14Cålder saknas i rapport Poz? 1460-...,NaN,Träkol,NaN,A3829,Stolphål,6302567,487812,14.800081,56.866713,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kadefors, Ola",2006.0,Boplatsen i Kvarteren Biskopshagen,"Smålands museum, Rapport 2006:51",Växjö
812,813,Växjö,Småland,Biskopshagen,Växjö 358,Grav- och boplatsområde,L1953:8423,NaN,UaOkänd64,NaN,NaN,NaN,c14_data_saknas,Labnr och 14Cålder saknas i rapport Ua? 1740-1...,J,Träkol,Salix sp,A583,Kokgrop,6302567,487812,14.800081,56.866713,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kadefors, Ola",2006.0,Boplatsen i Kvarteren Biskopshagen,"Smålands museum, Rapport 2006:51",Växjö
813,814,Växjö,Småland,Biskopshagen,Växjö 358,Grav- och boplatsområde,L1953:8423,NaN,UaOkänd65,NaN,NaN,NaN,c14_data_saknas,Labnr och 14Cålder sknas i rapport Ua? 760-400BC,J,Träkol,Al,A354,Härdgrop,6302567,487812,14.800081,56.866713,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kadefors, Ola",2006.0,Boplatsen i Kvarteren Biskopshagen,"Smålands museum, Rapport 2006:51",Växjö
814,815,Växjö,Småland,Biskopshagen,Växjö 358,Grav- och boplatsområde,L1953:8423,NaN,UaOkänd66,NaN,NaN,NaN,c14_data_saknas,Labnr och 14Cålder saknas i rapport 390-160 BC,J,Träkol,Al,A273,Härd,6302567,487812,14.800081,56.866713,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Kadefors, Ola",2006.0,Boplatsen i Kvarteren Biskopshagen,"Smålands museum, Rapport 2006:51",Växjö
1492,1493,Sunne,Värmland,Gylleby 1:11,Sunne 216,Bytomt/gårdstomt,L2006:8990,200300137.0,Ua-22050,NaN,NaN,NaN,c14_data_saknas,14C-åler saknas i rapport Kal 2 sigma 400-200 ...,J,Träkol,NaN,A86,Härd,6633692,396941,13.161699,59.828057,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Olsson, Hans",2004.0,Arkeologisk förundersökning i Torvnäsområdet,Värmlands museum 2004:14,Karlstad
1493,1494,Sunne,Värmland,Torvnäs 1:11,Sunne 329,Boplatsområde,L2005:6495,200300137.0,Ua-22046,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i rapport Kal 2 sigma 101-820...,J,Träkol,NaN,A75,Härd,6633993,396706,13.157361,59.830700,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Olsson, Hans",2004.0,Arkeologisk förundersökning i Torvnäsområdet,Värmlands museum 2004:14,Karlstad
1494,1495,Sunne,Värmland,Torvnäs 1:11,Sunne 330,Boplatsområde,L2005:645,200300137.0,Ua-22045,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i rapport Kal 2 sigma 1080-83...,J,Träkol,NaN,A74,Härd,6730401,361347,12.461079,60.685177,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Olsson, Hans",2004.0,Arkeologisk förundersökning i Torvnäsområdet,Värmlands museum 2004:14,Karlstad
1495,1496,Sunne,Värmland,Häljeby 1:58,Sunne 333,Boplatsområde,L2005:6454,200300137.0,Ua-22049,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i rapport Kal 2 sigma 1210-89...,J,Träkol,NaN,A32,Härd,6633460,396794,13.159194,59.825938,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Olsson, Hans",2004.0,Arkeologisk förundersökning i Torvnäsområdet,Värmlands museum 2004:14,Karlstad
1496,1497,Sunne,Värmland,Torvnäs 1:11,Sunne 334,Bolpats,L2005:6504,200300137.0,Ua-22044,NaN,NaN,NaN,c14_data_saknas,14C-ålder saknas i rasapport Kal 2 sigma 400-2...,J,Träkol,NaN,A65,Härd,6634266,396801,13.158919,59.833173,archaeological site,NaN,NaN,NaN,NaN,NaN,NaN,not calibratable,"Olsson, Hans",2004.0,Arkeologisk förundersökning i Torvnäsområdet,Värmlands museum 2004:14,Karlstad
1497,1498,Sunne,Värmland,Torvnäs 1:11,Sunne 334,Bolpats,L2005:

In [12]:
df_c14_data_saknas.to_csv('output/strucke_v8_no_c14_data.csv')

In [13]:
df[df['raa_id'].isna() & df['site_id'].isna()]

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,pMC_value,pMC_error,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
1491,1492,Mosjö,Närke,Marieberg,NaN,NaN,NaN,NaN,Ua-9615,390.0,70.0,25.11,NaN,NaN,ok,Antagligen felaktigt fornlämningsnummer angive...,J,Träkol,Tall,NaN,Grop,6563052,507348,15.128669,59.206488,parish,1628.0,1444.0,1645.0,1422.0,1522.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Pettersson, Olof",1998.0,Fjärrvärmeledning mellan Örebro och Kumla,Riksantikvarieämbetet Arkeologiska undersöknin...,Stockholm
1596,1597,Gällivare,Lappland,Kronoöverloppsmark 2:1 Lappviken,NaN,NaN,NaN,NaN,ST11446,905.0,190.0,27.40,NaN,NaN,ok,Saknas i Fornsök Lappviken,J,Träkol,NaN,A197,Härd,7472534,720188,20.115143,67.287762,parish,1284.0,900.0,1407.0,706.0,1106.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
1597,1598,Gällivare,Lappland,Kronoöverloppsmark 2:1 Lappviken,NaN,NaN,NaN,NaN,ST11448,990.0,70.0,27.30,NaN,NaN,ok,Saknas i Fornsök Lappviken,J,Träkol,NaN,A197,Härd,7472534,720188,20.115143,67.287762,parish,1156.0,994.0,1216.0,896.0,1074.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
1598,1599,Gällivare,Lappland,Kronoöverloppsmark 2:1 Luotosuollo,NaN,Viste,NaN,NaN,ST1851,875.0,65.0,0.00,NaN,NaN,ok,Oklart vilken fornlämning och position. Inlämn...,L,Träkol,NaN,1375:10,Härd,7472534,720188,20.115143,67.287762,parish,1254.0,1048.0,1266.0,1037.0,1165.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
1599,1600,Gällivare,Lappland,Kronoöverloppsmark 2:1 Luotosuollo,NaN,Viste,NaN,NaN,ST1852,810.0,65.0,0.00,NaN,NaN,ok,Oklart vilken fornlämning och position. Inlämn...,J,Träkol,NaN,1375:15,Stensättning,7472534,720188,20.115143,67.287762,parish,1276.0,1176.0,1298.0,1045.0,1222.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
28990,28991,Uppsala,Uppland,Ulls väg 31,NaN,Kokgrop,NaN,201300122.0,Ua-42008,1810.0,30.0,23.40,NaN,NaN,ok,Saknas i Fornsök. Ca 90 m V om L1940:4411,J,Träkol,En,A205,Kokgrop,6639689,647500,17.634433,59.868401,parish,317.0,213.0,336.0,131.0,242.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Hulth, Helena & Andersson, Fredrk",213.0,Förhistoriska utsnitt i Ultuna. Arkeologiska s...,SAU Rapport 2013:9,Uppsala
28991,28992,Uppsala,Uppland,Ulltuna 2:23,NaN,Grop,NaN,201300122.0,Ua-42010,3188.0,48.0,28.30,NaN,NaN,ok,Saknas i fornsök belägen ca 27 m NV om L1940:2160,J,Träkol,Ek,A1147,Grop,6639689,647500,17.634433,59.868401,parish,-1422.0,-1500.0,-1306.0,-1596.0,-1458.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Hulth, Helena & Andersson, Fredrk",213.0,Förhistoriska utsnitt i Ultuna. Arkeologiska s...,SAU Rapport 2013:9,Uppsala
29186,29187,Uppsala,Uppland,Emblas väg,NaN,Boplats,NaN,NaN,Ua-5379,1640.0,60.0,0.00,NaN,NaN,ok,Daterar ej grav utan snarare lager.,J,Förkolnat frö,Emmervete,NaN,Lager,6639689,647500,17.634433,59.868401,parish,537.0,366.0,564.0,255.0,433.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Roslund-Forenius, Ylva",1996.0,Gamla Uppsala,"UV Uppsala, Rapport 1996:39",Uppsala
29188,29189,Uppsala,Uppland,Vattholmavägen,NaN,Boplats,NaN,NaN,Ua-5381,1390.0,60.0,0.00,NaN,NaN,ok,Oklart läge för anläggningen,J,Träkol,NaN,NaN,NaN,6639689,647500,17.634433,59.868401,parish,770.0,597.0,773.0,561.0,647.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Roslund-Forenius, Ylva",1996.0,Gamla Uppsala,"

# Author full

In SEAD, tbl_biblio has the column named author whose values seem to be the author followed by the year in parenthesis

## Approach

`author` has 2020 distinct raw values with wildly inconsistent formatting: most are already `Surname, Firstname`, some are `Firstname Surname` with no comma at all, several authors get joined with `&` (or plain commas, or the Swedish word `och`), and on top of that there are typos, OCR-looking artifacts, `m fl` ("et al."), `(red)` ("ed."), and nobility particles (`von`, `von der`, ...) scattered around.

Since there are only ~2020 unique strings behind the 30k+ rows, the plan is:
1. Parse every unique author string once with a heuristic normalizer that targets the `Surname, N., Surname2, N2. & Surname3, N3.` pattern.
2. Flag anything the heuristic can't confidently reorder (rather than silently guessing) so it can be reviewed/corrected by hand.
3. Merge the normalized names back onto the full dataframe and append `(year)` to build `author_full`.

In [ ]:
unique_authors = sorted(df['author'].dropna().unique())
len(unique_authors)

## Step 1 — Heuristic parser

Rules, in order:
- Strip trailing `m fl` / `m. fl.` (→ " et al." suffix) and `(red)` (→ " (ed.)" suffix).
- Replace the Swedish conjunction `och` with `&`.
- If the string has commas: a single `&` always precedes the last author in this dataset, so replacing `&` with `,` and splitting on commas turns the whole thing into a flat token list. Since every author is written as exactly two tokens (`Surname, Firstname`), pairing tokens two-by-two recovers each author. An odd leftover token is checked once more in case it hides an unseparated `Surname Initial` pair (e.g. `"...,  Wallebom U"`).
- If the string has commas trailing a nobility particle (`von`, `von der`, `af`, ...) stuck to the wrong side of the pair, move the particle back onto the surname (`"Luft, Magnus von der"` → `"von der Luft, Magnus"`).
- If the string has **no comma at all**, name order is genuinely ambiguous (`"Firstname Surname"` vs `"Surname Firstname"`) — only two narrow, safe patterns are auto-fixed: a stray `.`/`;`/`:` typo'd in place of a comma (`"Edring. Anders"`), and a bare `Surname Initial` with no punctuation (`"Ebbesen K"`). Anything else with no comma is left untouched and flagged.
- First names are reduced to a single initial (`"Mattias"` → `"M."`, `"Anna-Karin"` → `"A."`).
- Anything with digits embedded in the name, or an odd token count that can't be repaired, is flagged `needs_review` and falls back to the original (suffix-stripped) text instead of a guess.

In [ ]:
import re

ET_AL_RE = re.compile(r'[,\s]*\(?\bm\.?\s*f\.?l\.?\)?\.?\s*$', re.I)
RED_RE = re.compile(r'[,\s]*\(?\s*red\.?\s*\)?\.?\s*$', re.I)
PARTICLE_RE = re.compile(r'\b(von|van|af|de)\b', re.I)
DIGIT_RE = re.compile(r'\d')
OCH_RE = re.compile(r'\s+och\s+', re.I)
NAME_WORD = r"[A-ZÅÄÖÜ][\wÀ-ÿ\-']*"
# "Surname. Firstname" / "Surname; Firstname" / "Surname: Firstname" (typo'd comma)
PSEUDO_COMMA_RE = re.compile(rf'^({NAME_WORD})[.;:]\s+({NAME_WORD})$')
# "Surname Initial" with no separator at all, e.g. "Ebbesen K" / "Adamsen Chr."
SHORT_INITIAL_RE = re.compile(rf'^({NAME_WORD})\s+([A-ZÅÄÖÜ][a-zåäöA-ZÅÄÖ]{{0,3}}\.?)$')
PARTICLE_WORDS = {'von', 'van', 'af', 'de', 'der', 'den', 'zu', 'du', 'la', 'le'}


def strip_suffixes(s):
    """Pull off trailing 'm fl' (et al.) / '(red)' (ed.) annotations, in either order."""
    et_al = False
    editor = False
    changed = True
    while changed:
        changed = False
        new_s = ET_AL_RE.sub('', s)
        if new_s != s:
            et_al, s, changed = True, new_s, True
        new_s = RED_RE.sub('', s)
        if new_s != s:
            editor, s, changed = True, new_s, True
    return s.strip(' ,.'), et_al, editor


def fix_particle_order(surname, firstname):
    """Move a trailing nobility particle (e.g. 'Magnus von der') from the firstname
    back onto the surname ('von der Luft'), which is where OCR/data-entry usually lost it."""
    words = firstname.split()
    particles = []
    while words and words[-1].lower() in PARTICLE_WORDS:
        particles.insert(0, words.pop().lower())
    if particles:
        surname = ' '.join(particles) + ' ' + surname
        firstname = ' '.join(words)
    return surname, firstname


def initialize(name_part):
    """Turn a first-name segment into an initial, e.g. 'Mattias' -> 'M.', 'Anna-Karin' -> 'A.'"""
    name_part = name_part.strip(' .')
    if not name_part:
        return ''
    first_word = re.split(r'[\s\-]', name_part)[0]
    return first_word[0].upper() + '.' if first_word else ''


def parse_no_comma_segment(segment):
    """Try to safely split a single no-comma author segment into (surname, firstname).
    Returns None if the order can't be determined with confidence."""
    segment = segment.strip()
    m = PSEUDO_COMMA_RE.match(segment) or SHORT_INITIAL_RE.match(segment)
    return (m.group(1), m.group(2)) if m else None


def build_result(original, formatted, et_al, editor, flags):
    suffix = ''
    if et_al:
        suffix += ' et al.'
    if editor:
        suffix += ' (ed.)'
    formatted_full = (formatted + suffix).strip()
    needs_review = 'needs_review' in flags
    reason = ','.join(sorted(set(f for f in flags if f != 'needs_review')))
    return original, formatted_full, needs_review, reason


def parse_author_string(raw):
    original = raw
    flags = []

    s = OCH_RE.sub(' & ', raw.strip())
    s, et_al, editor = strip_suffixes(s)
    if et_al:
        flags.append('et_al')
    if editor:
        flags.append('editor')
    if DIGIT_RE.search(s):
        flags.append('digits')
    if PARTICLE_RE.search(s):
        flags.append('particle')

    if ',' not in s:
        flags.append('no_comma')
        segments = [seg.strip() for seg in re.split(r'\s*&\s*', s) if seg.strip()]
        parsed_pairs = [parse_no_comma_segment(seg) for seg in segments]
        if all(parsed_pairs):
            authors = [f'{surname}, {initialize(firstname)}'.strip() for surname, firstname in parsed_pairs]
            formatted = authors[0] if len(authors) == 1 else ', '.join(authors[:-1]) + ' & ' + authors[-1]
        else:
            formatted = s  # can't safely determine name order
            flags.append('needs_review')
        return build_result(original, formatted, et_al, editor, flags)

    # a single '&' always precedes the final author in this dataset (verified: no multi-'&' cases)
    s_norm = re.sub(r'\s*&\s*', ', ', s)
    tokens = [t.strip() for t in s_norm.split(',') if t.strip()]

    if len(tokens) % 2 != 0:
        # repair pass: the last token may itself hide an unseparated "Surname Initial" pair
        m = SHORT_INITIAL_RE.match(tokens[-1]) if tokens else None
        if m:
            tokens = tokens[:-1] + [m.group(1), m.group(2)]

    if len(tokens) % 2 != 0:
        flags += ['odd_tokens', 'needs_review']
        return build_result(original, s, et_al, editor, flags)

    authors = []
    for i in range(0, len(tokens), 2):
        surname, firstname = tokens[i], tokens[i + 1]
        if 'particle' in flags:
            surname, firstname = fix_particle_order(surname, firstname)
        initial = initialize(firstname)
        if not initial:
            flags.append('needs_review')
        authors.append(f'{surname}, {initial}'.strip())

    formatted = authors[0] if len(authors) == 1 else ', '.join(authors[:-1]) + ' & ' + authors[-1]
    if 'digits' in flags:
        flags.append('needs_review')

    return build_result(original, formatted, et_al, editor, flags)

## Step 2 — Run the parser over every unique value and flag uncertain ones

The result is saved to `output/author_normalization_review.csv` — a lookup table with one row per unique raw `author` value, its parsed form, and whether it needs a human look. Rows flagged `needs_review` fall back to the original (suffix-stripped) text rather than a guessed reordering, so they're easy to spot and hand-correct in the CSV later.

In [ ]:
rows = []
for a in unique_authors:
    original, formatted, needs_review, reason = parse_author_string(a)
    rows.append({'author_raw': original, 'author_parsed': formatted, 'needs_review': needs_review, 'flag_reason': reason})

author_review_df = pd.DataFrame(rows)
author_review_df.to_csv('output/author_normalization_review.csv', index=False, encoding='utf-8-sig')

total = len(author_review_df)
flagged = author_review_df['needs_review'].sum()
print(f'total unique authors: {total}')
print(f'flagged for review: {flagged} ({flagged/total:.1%})')
print()
print(author_review_df.loc[author_review_df['needs_review'], 'flag_reason'].value_counts())

A peek at what got flagged and why — mostly genuinely ambiguous name order (`no_comma`), malformed comma counts (`odd_tokens`), or garbage/OCR artifacts (`digits`):

In [ ]:
author_review_df[author_review_df['needs_review']].sample(15, random_state=1)[['author_raw', 'author_parsed', 'flag_reason']]

## Step 3 — Merge the normalized names back onto `df` and add the year

`author_review_df` is a one-row-per-unique-value lookup, so merging it onto `df` on the `author` column is a clean left join with no row duplication. `author_full` then appends `(publication_year)` — left off entirely when the year is missing rather than producing an empty `()`.

In [ ]:
df_new = df.merge(
    author_review_df[['author_raw', 'author_parsed', 'needs_review']],
    left_on='author', right_on='author_raw', how='left'
).drop(columns='author_raw').rename(columns={'author_parsed': 'author_normalized', 'needs_review': 'author_needs_review'})

df_new['author_full'] = df_new.apply(
    lambda r: f"{r['author_normalized']} ({int(r['publication_year'])})" if pd.notna(r['publication_year']) else r['author_normalized'],
    axis=1
)

print(f"rows flagged for manual author review: {df_new['author_needs_review'].sum()} / {len(df_new)}")
df_new[['author', 'author_normalized', 'author_needs_review', 'publication_year', 'author_full']]

## Result

`df_new['author_full']` now holds `Surname, N., Surname2, N2. & Surname3, N3. (year)`, built from ~2020 unique author strings normalized once and mapped back onto all 30,301 rows.

~4% of rows (`author_needs_review == True`) carry a raw author string the heuristic wasn't confident about (ambiguous `Firstname Surname` order, malformed punctuation, or OCR-looking garbage) — for those, `author_normalized` falls back to the original text. To fix them: edit `author_parsed` in `output/author_normalization_review.csv` for the flagged rows, then re-run the Step 3 merge cell to refresh `df_new`.

# Unique (material, species) combinations

Extract every distinct `(material, species)` pair present in the data, with how many rows each pair covers.

In [4]:
material_species = (
    df[['material', 'species']]
    .value_counts(dropna=False)
    .reset_index(name='n_rows')
    .sort_values(['material', 'species'], na_position='first')
    .reset_index(drop=True)
)

print(f'{len(material_species)} unique (material, species) tuples')
material_species

755 unique (material, species) tuples


,material,species,n_rows
0,NaN,NaN,1
1,Animalier,Fiskfjäll,1
2,Animalier,Hjorthår,1
3,Animalier,"Hjärnsubstans, människa",1
4,Animalier,Hår,1
...,...,...,...
750,Växtdelar,Örtstam och rot,1
751,Växtdelar,Örtstjälk,1
752,Växtdelar,Örtstjälkar,1
753,Växtdelar,ört,1


In [6]:
material_species.to_csv('output/strucke_v8_material_species_unique.csv', index=False, encoding='utf-8-sig')

#### Rows with no real information of the site (site_id as the only real information about the site since the column raa_id is often not enough)

In [5]:
df[df['site_id'].isna()]

,fid,socken,landskap,village_farm,raa_id,site_type,site_id,uppdragsnummer,lab_no,c14_age_bp,c14_error,d13C,pMC_value,pMC_error,c14_data_status,comment,assessed_relevant,material,species,context_id,context_type,northing_3006,easting_3006,longitude,latitude,location_precision,cal_68_min,cal_68_max,cal_95_min,cal_95_max,median_cal_year,calibration_method,calibration_status,author,publication_year,title,journal,place_of_publication
1491,1492,Mosjö,Närke,Marieberg,NaN,NaN,NaN,NaN,Ua-9615,390.0,70.0,25.11,NaN,NaN,ok,Antagligen felaktigt fornlämningsnummer angive...,J,Träkol,Tall,NaN,Grop,6563052,507348,15.128669,59.206488,parish,1628.0,1444.0,1645.0,1422.0,1522.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Pettersson, Olof",1998.0,Fjärrvärmeledning mellan Örebro och Kumla,Riksantikvarieämbetet Arkeologiska undersöknin...,Stockholm
1596,1597,Gällivare,Lappland,Kronoöverloppsmark 2:1 Lappviken,NaN,NaN,NaN,NaN,ST11446,905.0,190.0,27.40,NaN,NaN,ok,Saknas i Fornsök Lappviken,J,Träkol,NaN,A197,Härd,7472534,720188,20.115143,67.287762,parish,1284.0,900.0,1407.0,706.0,1106.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
1597,1598,Gällivare,Lappland,Kronoöverloppsmark 2:1 Lappviken,NaN,NaN,NaN,NaN,ST11448,990.0,70.0,27.30,NaN,NaN,ok,Saknas i Fornsök Lappviken,J,Träkol,NaN,A197,Härd,7472534,720188,20.115143,67.287762,parish,1156.0,994.0,1216.0,896.0,1074.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
1598,1599,Gällivare,Lappland,Kronoöverloppsmark 2:1 Luotosuollo,NaN,Viste,NaN,NaN,ST1851,875.0,65.0,0.00,NaN,NaN,ok,Oklart vilken fornlämning och position. Inlämn...,L,Träkol,NaN,1375:10,Härd,7472534,720188,20.115143,67.287762,parish,1254.0,1048.0,1266.0,1037.0,1165.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
1599,1600,Gällivare,Lappland,Kronoöverloppsmark 2:1 Luotosuollo,NaN,Viste,NaN,NaN,ST1852,810.0,65.0,0.00,NaN,NaN,ok,Oklart vilken fornlämning och position. Inlämn...,J,Träkol,NaN,1375:15,Stensättning,7472534,720188,20.115143,67.287762,parish,1276.0,1176.0,1298.0,1045.0,1222.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Mulk, Inga-Maria",1994.0,Sirkas,Studia archaeologica Universitatis Umensis 6,Umeå
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
28990,28991,Uppsala,Uppland,Ulls väg 31,NaN,Kokgrop,NaN,201300122.0,Ua-42008,1810.0,30.0,23.40,NaN,NaN,ok,Saknas i Fornsök. Ca 90 m V om L1940:4411,J,Träkol,En,A205,Kokgrop,6639689,647500,17.634433,59.868401,parish,317.0,213.0,336.0,131.0,242.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Hulth, Helena & Andersson, Fredrk",213.0,Förhistoriska utsnitt i Ultuna. Arkeologiska s...,SAU Rapport 2013:9,Uppsala
28991,28992,Uppsala,Uppland,Ulltuna 2:23,NaN,Grop,NaN,201300122.0,Ua-42010,3188.0,48.0,28.30,NaN,NaN,ok,Saknas i fornsök belägen ca 27 m NV om L1940:2160,J,Träkol,Ek,A1147,Grop,6639689,647500,17.634433,59.868401,parish,-1422.0,-1500.0,-1306.0,-1596.0,-1458.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Hulth, Helena & Andersson, Fredrk",213.0,Förhistoriska utsnitt i Ultuna. Arkeologiska s...,SAU Rapport 2013:9,Uppsala
29186,29187,Uppsala,Uppland,Emblas väg,NaN,Boplats,NaN,NaN,Ua-5379,1640.0,60.0,0.00,NaN,NaN,ok,Daterar ej grav utan snarare lager.,J,Förkolnat frö,Emmervete,NaN,Lager,6639689,647500,17.634433,59.868401,parish,537.0,366.0,564.0,255.0,433.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Roslund-Forenius, Ylva",1996.0,Gamla Uppsala,"UV Uppsala, Rapport 1996:39",Uppsala
29188,29189,Uppsala,Uppland,Vattholmavägen,NaN,Boplats,NaN,NaN,Ua-5381,1390.0,60.0,0.00,NaN,NaN,ok,Oklart läge för anläggningen,J,Träkol,NaN,NaN,NaN,6639689,647500,17.634433,59.868401,parish,770.0,597.0,773.0,561.0,647.0,HPD; IntCal20; rcarbon 1.5.0,calibrated,"Roslund-Forenius, Ylva",1996.0,Gamla Uppsala,"